src/impl.cpp

```diff
@@ -0,0 +1,16 @@
+#include "impl.h"
+
+int BinaryInvert(int x,int p,int n)
+{
+    constexpr int INT32_BIT_COUNT = sizeof(x)*8;
+
+    if (p < INT32_BIT_COUNT && n <= p)
+    {
+        for (int i = p; i > p - n; i--)
```

Comment:
Yep, that's right. Also, you can prepare the bitmask in two lines with shifts and then just apply it.